### ISSUE TITLE
[prefer-readonly-parameter-types] always fails on TypedArray parameters

### ISSUE THREAD

**Issue by bitjson** [typescript-eslint/typescript-eslint#1668]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": ["warn"]
  }
}
```

```TS
const firstByte = (bin: Readonly<Uint8Array>) => bin[0];
```

**Expected Result**
Should not report, since `Uint8Array` has been wrapped in the `Readonly` Utility type. 

(I think this is a strong guarantee of the `Uint8Array` being readonly. Even when attempting to reassign values of specific indexes, Typescript throws an error: `Index signature in type 'Readonly<Uint8Array>' only permits reading.`)

**Actual Result**
Even with the `Readonly<Uint8Array>`, the rule still warns: `Parameter should be a read only type`

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `2.22.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.1` |
| `npm`                              | `6.14.2` |

**Comment by bradzacher** [comment]
This is being flagged because the type is not recursively readonly.
Specifically, the `buffer` property, which is an `ArrayBuffer`, is not readonly.

I.e. this is compile-time valid typescript:
```ts
const x: Readonly<Uint8Array> = new Uint8Array(0);
x.buffer.slice = () => {
    console.log('boop');
    return new ArrayBuffer(0);
};

x.buffer.slice(0);
// -> 'boop'
```

Hmmm. I'm not sure the best solution here, because:
- it was an intentional decision to not treat a type as readonly unless it was recursively readonly.
    - you don't want to pass in an object with nested objects then accidentally mutate the nested object.
- it was an intentional decision to not treat methods as readonly
    - it's possible to reassign methods, as they are not readonly by default.

I'm not sure of a good strategy to do recursively readonly types.
There are types in the ecosystem like those provided by the [`type-fest`](https://github.com/sindresorhus/type-fest/blob/master/source/readonly-deep.d.ts) package which make it easier to do deep readonly types:

```ts
import { ReadonlyDeep } from 'type-fest';

const x: ReadonlyDeep<Uint8Array> = new Uint8Array(0);
x.buffer.slice = () => { // Error: Cannot assign to 'slice' because it is a read-only property.
    console.log('boop');
    return new ArrayBuffer(0);
};
```
[repl](https://www.typescriptlang.org/play/?ssl=10&ssc=52&pln=10&pc=61#code/AQFwngDgpsAKBOBLAtokiBuMC8AoAkAD7AB2ArgDYUHFkkAmUAZoiVPTcAM4hIkDmncsgBGUeJxEB7KRSgBDEpy5hRsyYn6sQAbly5w0YACUF9KSQpgAIlCgQAPABUAfMGzAnwKAA8QUBi5gAAoEFDRMGGJg4IA6ePl4fgAuYEUwAG0AXQBKdzc6AGsSKQB3EhycggB+TwJUr19-QJMzCysAWXlHViZxYABpKDAnSCgAGmBe-oA1eQoyKFHoFxrW+XNLMC6IW3sHIZGxybmFpbHV-AbvPwD6INMN9rAAZSgQB2n4YABJf2RllBLrVHpsrG8QHtHH8oACLvVPDdmvdgFIRAArKAAYxAa1BzwA8hjsZC7I5XLhgFTgKkiiVynpcAB6ABULNwL3kyBg8iCAAN8VsoXzJiIyCA0lisfYQEFnsABW0tjs+UFeVMSBBxVxYr8SP54CR5sAABZQCjQb5MKTfRVPIVkvmxXAspm4bTiJjyaXrMHbbpQg7DQEneaLQGXJp3B5KzrdByCqyBw4RyaJmxkhyncMXNwAbwAvvpWezOdy0vz08LReLJdKILLUVsFemIaqKxqtbLdT99eIjRRTebLcBrbaq47na73X34F6fa33oGYXCVgQoy1Fx8J-sVxH80XmWyOVyeZXYxn7CLgGKJd76435Xy0Zice31awuzq9QaB0OLf0Y4thewpTm6hgwOmRKvqS+zQSSgJItGqLEjibgeHmBDwBewAZCmYwasAhTDFITDAPBOKAlkqQ7o4FEgICeHBmMWSrAWjJYhYPDAD4NEgZmACq2gABwAILwPA8hgOhpBQKUwBCfqYkSVJwQAAw5HoPixGKTB9PAsRcBQiA+h4wR5Ng+aUtSnEkFwshQLEFBSPwwQAOTSFIEBuZp1lUthIBkIasnyeJklgAAQmQeniOpvnsfo2m6fphnGdKcU6EAA).

Similarly there is this issue https://github.com/microsoft/TypeScript/issues/13923 to get a deep readonly type added to TS's base lib.

**Comment by bitjson** [comment]
Ah, thanks for the detailed response! I was mistaken, `Readonly<Uint8Array>` is not enough, and a deep-readonly utility type is a good solution. I'm using one listed in the [GitHub issue you linked](https://github.com/microsoft/TypeScript/issues/13923#issuecomment-557509399). Thanks!

I don't think any changes need to be made to the rule, TypeScript just needs a built-in `DeepReadonly`. 

If anything, it might be helpful to add a note about this to the documentation for `prefer-readonly-parameter-types`.
